Further reduce use of std::span::subspan() by leveraging ParsingUtilities

cdumez opened the pull request:
#### 4e98b4d7345d77ada364888535ef651e7dc53f70
<pre>
Further reduce use of std::span::subspan() by leveraging ParsingUtilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=285283">https://bugs.webkit.org/show_bug.cgi?id=285283</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/DateMath.cpp:
(WTF::parseDate):
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::mapToFile):
* Source/WTF/wtf/JSONValues.cpp:
* Source/WTF/wtf/StreamBuffer.h:
(WTF::StreamBuffer::append):
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::userVisibleURL):
* Source/WTF/wtf/text/Base64.cpp:
(WTF::fromBase64SlowImpl):
* Source/WTF/wtf/text/ParsingUtilities.h:
(WTF::skipUntil):
(WTF::skipWhile):
(WTF::consumeSpan):
(WTF::consumeSingleElementOfType):
* Source/WTF/wtf/text/StringBuilderJSON.cpp:
(WTF::StringBuilder::appendQuotedJSONString):
* Source/WTF/wtf/text/StringBuilderJSON.h:
(WTF::appendEscapedJSONStringContent):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::simplifyMatchedCharactersToSpace):
* Source/WTF/wtf/text/StringToIntegerConversion.h:
(WTF::parseInteger):
* Source/WebCore/Modules/indexeddb/server/IDBSerialization.cpp:
(WebCore::readLittleEndian):
(WebCore::decodeKey):
(WebCore::deserializeIDBKeyData):
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformer.cpp:
(WebCore::parseSFrameHeader):
* Source/WebCore/Modules/mediastream/STUNMessageParsing.cpp:
(WebCore::WebRTC::extractDataMessages):
* Source/WebCore/Modules/websockets/WebSocketExtensionParser.cpp:
(WebCore::isSpaceOrTab):
(WebCore::WebSocketExtensionParser::skipSpaces):
(WebCore::WebSocketExtensionParser::consumeToken):
(WebCore::WebSocketExtensionParser::consumeQuotedString):
(WebCore::WebSocketExtensionParser::consumeCharacter):
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
(PAL::TextCodecLatin1::decode):
* Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp:
(PAL::TextCodecUTF8::handlePartialSequence):
(PAL::TextCodecUTF8::decode):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readLittleEndian):
(WebCore::CloneDeserializer::readTag):
(WebCore::CloneDeserializer::readArrayBufferViewSubtag):
(WebCore::CloneDeserializer::readResizableNonSharedArrayBuffer):
(WebCore::CloneDeserializer::read):
* Source/WebCore/css/CSSMarkup.cpp:
(WebCore::isCSSTokenizerIdentifier):
* Source/WebCore/css/CSSStyleDeclaration.cpp:
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseColorIntOrPercentage):
(WebCore::parseRGBAlphaValue):
(WebCore::parseLegacyHSL):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::SearchBuffer::isBadMatch const):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(consume):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::isHTMLSpaceOrDelimiterOrNumberStart):
(WebCore::parseHTMLListOfOfFloatingPointNumberValuesInternal):
(WebCore::isASCIIDigitOrPeriod):
(WebCore::parseHTTPRefreshInternal):
* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::TextResourceDecoder::checkForCSSCharset):
(WebCore::TextResourceDecoder::checkForHeadCharset):
* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::SharedMemory::copyBuffer):
* Source/WebCore/platform/audio/SincResampler.cpp:
(WebCore::SincResampler::processBuffer):
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseHTTPHeader):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::popDescriptorAndAdvance):
(IPC::Connection::sendOutgoingMessage):
(IPC::createMessageDecoder):
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataDecoder::operator&gt;&gt;):
(WebKit::HistoryEntryDataDecoder::decodeFixedLengthData):

Canonical link: <a href="https://commits.webkit.org/288387@main">https://commits.webkit.org/288387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3b6c3e9835ef620976b4768f892792bbd7dd6e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88061 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33998 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2667 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10510 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64628 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/22391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86003 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1968 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75450 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44904 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1876 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29643 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33032 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75933 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89427 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81996 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73050 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10462 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71258 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72270 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16435 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15189 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12833 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10187 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104400 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10059 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25287 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11827 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/URLHelpers.cpp:
```diff
@@ -941,14 +942,13 @@ String userVisibleURL(const CString& url)
         while (p.front()) {
-            unsigned char c = p.front();
+            unsigned char c = consumeSingleElement(p);
             if (c > 0x7f) {
```
This should use `isASCII`

darinadler commented on Source/WTF/wtf/text/ParsingUtilities.h:
```diff
@@ -304,6 +321,19 @@ template<typename T> std::span<T> consumeSpan(std::span<T>& data, size_t amountT
+}
+
+template<typename DestinationType, typename SourceType>
+match_constness_t<SourceType, DestinationType>& consumeSingleElementOfType(std::span<SourceType>& data) requires(sizeof(SourceType) == 1)
```
This name is long. It also seems kind of related to `byteCast`, but doesn't mention byte.

cdumez replied:
I wanted "consume" in the name because it consumes part of the span. We already `consumeSingleElement()` (which I feel is kind of long as well and maybe we could rename to just `consume()`). This one consumes one element of a given type from a span of bytes so I just added "OfType" to the name.

Maybe `consumeAndCastTo<>()` ?

cdumez replied:
I went with `consumeAndCastTo<>()`, let me know if you have a better suggestion.

darinadler replied:
I'm not sure it's better but I would name it `consumeByte`.

cdumez replied:
But it doesn't consume a byte so I'd find it confusing. It is used as follows:
```
ObjectType element = consumeAndCastTo<ObjectType>(byteSpan);
```

It consumes `sizeof(ObjectType)` bytes in `byteSpan` and cast those bytes to `ObjectType`.

darinadler replied:
Oh I misread the "size must be 1" part of the code

darinadler replied:
I like the name you chose but also think `consumeCast<>` would be good.

darinadler commented on Source/WebCore/css/CSSStyleDeclaration.cpp:
```diff
@@ -173,10 +160,8 @@ static CSSPropertyID parseJavaScriptCSSPropertyName(const AtomString& propertyNa
             bufferSpan[0] = '-';
             bufferSpan[1] = toASCIILowerUnchecked(c);
             bufferSpan = bufferSpan.subspan(2);
```
Seems like a nice idiom for this kind of thing when it can't be merged with the code above would be something like `skip(bufferSpan, 2)`

cdumez replied:
OK, will add a `skip()` function. It will make the code a bit more concise.

darinadler commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1462,9 +1463,7 @@ static Class _WebMessageDocumentClass()
-    span = span.subspan(1);
-    return character;
+    return consumeSingleElement(span);
 }
```
Seems funny to wrap this just to shorten the function name. Maybe `consumeSingleElement` can just be named `consume`.

cdumez replied:
Yes, I was hesitating. I'll rename `consumeSingleElement` to `consume`

cdumez committed:
Further reduce use of std::span::subspan() by leveraging ParsingUtilities
https://bugs.webkit.org/show_bug.cgi?id=285283

Reviewed by Darin Adler.

* Source/WTF/wtf/DateMath.cpp:
(WTF::parseDate):
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::mapToFile):
* Source/WTF/wtf/JSONValues.cpp:
* Source/WTF/wtf/StreamBuffer.h:
(WTF::StreamBuffer::append):
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::userVisibleURL):
* Source/WTF/wtf/text/Base64.cpp:
(WTF::fromBase64SlowImpl):
* Source/WTF/wtf/text/ParsingUtilities.h:
(WTF::skipUntil):
(WTF::skipWhile):
(WTF::consumeSpan):
(WTF::consumeSingleElementOfType):
* Source/WTF/wtf/text/StringBuilderJSON.cpp:
(WTF::StringBuilder::appendQuotedJSONString):
* Source/WTF/wtf/text/StringBuilderJSON.h:
(WTF::appendEscapedJSONStringContent):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::simplifyMatchedCharactersToSpace):
* Source/WTF/wtf/text/StringToIntegerConversion.h:
(WTF::parseInteger):
* Source/WebCore/Modules/indexeddb/server/IDBSerialization.cpp:
(WebCore::readLittleEndian):
(WebCore::decodeKey):
(WebCore::deserializeIDBKeyData):
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformer.cpp:
(WebCore::parseSFrameHeader):
* Source/WebCore/Modules/mediastream/STUNMessageParsing.cpp:
(WebCore::WebRTC::extractDataMessages):
* Source/WebCore/Modules/websockets/WebSocketExtensionParser.cpp:
(WebCore::isSpaceOrTab):
(WebCore::WebSocketExtensionParser::skipSpaces):
(WebCore::WebSocketExtensionParser::consumeToken):
(WebCore::WebSocketExtensionParser::consumeQuotedString):
(WebCore::WebSocketExtensionParser::consumeCharacter):
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
(PAL::TextCodecLatin1::decode):
* Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp:
(PAL::TextCodecUTF8::handlePartialSequence):
(PAL::TextCodecUTF8::decode):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readLittleEndian):
(WebCore::CloneDeserializer::readTag):
(WebCore::CloneDeserializer::readArrayBufferViewSubtag):
(WebCore::CloneDeserializer::readResizableNonSharedArrayBuffer):
(WebCore::CloneDeserializer::read):
* Source/WebCore/css/CSSMarkup.cpp:
(WebCore::isCSSTokenizerIdentifier):
* Source/WebCore/css/CSSStyleDeclaration.cpp:
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseColorIntOrPercentage):
(WebCore::parseRGBAlphaValue):
(WebCore::parseLegacyHSL):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::SearchBuffer::isBadMatch const):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(consume):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::isHTMLSpaceOrDelimiterOrNumberStart):
(WebCore::parseHTMLListOfOfFloatingPointNumberValuesInternal):
(WebCore::isASCIIDigitOrPeriod):
(WebCore::parseHTTPRefreshInternal):
* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::TextResourceDecoder::checkForCSSCharset):
(WebCore::TextResourceDecoder::checkForHeadCharset):
* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::SharedMemory::copyBuffer):
* Source/WebCore/platform/audio/SincResampler.cpp:
(WebCore::SincResampler::processBuffer):
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseHTTPHeader):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::popDescriptorAndAdvance):
(IPC::Connection::sendOutgoingMessage):
(IPC::createMessageDecoder):
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataDecoder::operator>>):
(WebKit::HistoryEntryDataDecoder::decodeFixedLengthData):

Canonical link: https://commits.webkit.org/288387@main
